mlir/lib/Dialect/MemRef/Transforms/MergeAllocTickBased.cpp

```
@@ -0,0 +1,383 @@
+  llvm::DenseMap<Operation *, llvm::SmallVector<TraceWithTick, 8>> raw;
+  for (auto &[op, tick] : s->allocTicks) {
+    if (!isMergeableAlloc(s, op, tick.firstAccess)) {
+      continue;
+    }
+    auto scope = getAllocScope(s, op);
+    if (!scope) {
+      return op->emitError(
+          "This op should be surrounded by an AutomaticAllocationScope");
+    }
+    auto allocSize = getAllocSize(s, op);
+    if (failed(allocSize)) {
+      return failure();
+    }
+    // tick.firstAccess * 2 and tick.lastAccess * 2 + 1 to make sure "dealloc"
+    // overlaps "alloc"
+    raw[scope].emplace_back(tick.allocTick, tick.firstAccess * 2,
+                            reinterpret_cast<uintptr_t>(op), *allocSize);
+    raw[scope].emplace_back(tick.allocTick, tick.lastAccess * 2 + 1,
+                            reinterpret_cast<uintptr_t>(op), 0);
+  }
+  MemoryTraceScopes ret;
+  for (auto &[scope, trace] : raw) {
+    std::stable_sort(trace.begin(), trace.end(),
+                     [](const TraceWithTick &a, const TraceWithTick &b) {
+                       if (a.tick == b.tick) {
+                         return a.allocTick < b.allocTick;
+                       }
+                       return a.tick < b.tick;
+                     });
+    auto retTrace = std::make_unique<TickTraceResult>();
+    retTrace->traces.reserve(trace.size());
+    for (auto &tr : trace) {
+      retTrace->traces.emplace_back(tr.trace);
+    }
+    ret.scopeToTraces[scope] = std::move(retTrace);
+  }
+  return ret;
+}
+
+FailureOr<MemoryTraceScopes>
+TickCollecter::operator()(Operation *root,
+                          const mlir::BufferViewFlowAnalysis &aliasAnaly,
+                          const MergeAllocationOptions &option) const {
+  TickCollecterStates s{aliasAnaly, option};
+  TickCollecter collecter;
+  LogicalResult result = success();
+  root->walk<WalkOrder::PreOrder>([&](Operation *op) {
+    if (failed(collecter.popScopeIfNecessary(&s, op))) {
+      result = failure();
```

Can it just return `failure()` directly instead of lazy return?

---

Not sure if it can be done inside of `walk<...>()`

---

Seems we can simply use `return;` or `return WalkResult::interrupt();` to stop earlier.

---

Thanks for the suggestions. Changed now.